### DISCUSSION TITLE
Feature request: Custom snapshot serialization

### DISCUSSION THREAD

**Discussion by schmod** [avajs/ava#2817] [Ideas]
Would it be possible for Ava to implement some sort of mechanism for us to (globally or per-test) specify how snapshots should be serialized and stored?

Our APIs use Luxon extensively for handling dates/times.  Unfortunately, Ava serializes the _entire_ Luxon object when we attempt to snapshot an object that contains a Luxon `DateTime` value.  

Excerpt from a snapshot:
```
          updated_at: DateTime {
            1628802455000
            ---
            _zone: LocalZone {},
            c: {
              day: 12,
              hour: 17,
              millisecond: 0,
              minute: 7,
              month: 8,
              second: 35,
              year: 2021,
            },
            invalid: null,
            isLuxonDateTime: true,
            loc: Locale {
              eraCache: {},
              fastNumbersCached: null,
              intl: 'en-US',
              locale: 'en-US',
              meridiemCache: null,
              monthsCache: {
                format: {},
                standalone: {},
              },
              numberingSystem: null,
              outputCalendar: null,
              specifiedLocale: null,
              weekdaysCache: {
                format: {},
                standalone: {},
              },
            },
            o: -240,
            ts: 1628802455000,
            weekData: null,
          },
```

Ideally, it'd be nice if we could somehow tell Ava that it should serialize `DateTime` objects by calling `.toString()` instead of dumping the entire object into the snapshot. That way, the snapshot of this property would look something like:

```
          updated_at: '2021-08-20T14:26:08.387-04:00'
```

I'd be fine with some sort of "middleware" approach where we could specify a `snapshotSerializer` in Ava's config (or on a per-file basis by calling a hypothetical `test.snapshotSerializer()` method.  This would be a path to a script file that contains a function that Ava calls before it serializes an object.  From there, we could implement our own handling for any classes that would benefit from normalization or custom serialization.  

This could also allow some Ava users to strip/normalize fields/data that are expected to be random or based on a timestamp.  (This is especially common in integration tests where it isn't possible to control the timestamps generated by a database or external system)

**Comment by novemberborn** [comment]
Yes, see https://github.com/avajs/ava/issues/1861. This would also work for `t.deepEqual()`. The underlying system is extensible but we've not yet properly exposed this as an API.

> This could also allow some Ava users to strip/normalize fields/data that are expected to be random or based on a timestamp. (This is especially common in integration tests where it isn't possible to control the timestamps generated by a database or external system)

I use a helper pattern for that, either directly (`t.snapshot(normalizeSomeData(data))`) or indirectly (`snapshotSomeData(t, data)`).
